sw/ground_segment/cockpit/shapes.ml

```diff
@@ -0,0 +1,112 @@
+(*
+ * This file is part of paparazzi.
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ *
+ *)
+
+ module G = MapCanvas
+ open Latlong
+ module LL = Latlong
+ open Printf
+
+
+type shdata = {
+  shid : string;
+  shlinecolor : string;
+  shfillcolor : string;
+  shtype : float;
+  shstatus : float;
+  shlatarr : float array;
+  shlonarr : float array;
+  shradius : float}
+
+type circleshape = { mutable circsh : GnoCanvas.ellipse}
+type polygonshape = { mutable polysh : GnoCanvas.polygon}
+
+let circleshapes = Hashtbl.create 1
+let polygonshapes = Hashtbl.create 1
+
+let circle_exist = fun id ->
+  Hashtbl.mem circleshapes id
+
+let polygon_exist = fun id ->
+  Hashtbl.mem polygonshapes id
+
+let update_circle = fun id wgs84 color ?fill_color radius (geomap:MapCanvas.widget) ->
+  try
+    let gencircle = geomap#circle ~width:2 ?fill_color ~color:color wgs84.(0) radius in
+    if not (circle_exist id) then
+    let circleshape = {circsh = gencircle } in
+    Hashtbl.add circleshapes id circleshape;
+    else
+    let shape = Hashtbl.find circleshapes id in
+    shape.circsh#destroy ();
```

Comment:
I was having issues setting a new circle with  `#absolute_affinity`

`let update_circle = fun id wgs84 color ?fill_color radius (geomap:MapCanvas.widget) ->
  try
    let gencircle = geomap#circle ~width:2 ?fill_color ~color:color wgs84.(0) radius in
    if (circle_exist id) then
    let shape = Hashtbl.find circleshapes id in
    shape#affine_absolute (geomap#convert_positions_to_points wgs84);
    else
    Hashtbl.add circleshapes id gencircle
  with _ -> ()`

Code like this will generate a circle and allow the change in location the once. Subsequent calls just draw a new circle.